op-contracts: Add script to upgrade from 1.8.0 L2OO to 1.8.0 permissioned.

ajsutton opened the pull request:
**Description**

Adds an upgrade package to transition from 1.8.0 in L2OO configuration to 1.8.0 in permissioned configuration.  This is based on the 1.3.0->1.6.0-permissioned docker image with minor changes to handle deploying the correct version of contracts like `AnchorStateRegistry` which was not released as part of 1.8.0, so needs to be deployed via a blueprint of the 1.6.0 implementation.

A runbook using this is added in https://github.com/ethereum-optimism/superchain-ops/pull/393

**Metadata**

Part of https://github.com/ethereum-optimism/optimism/issues/12990

ajsutton committed:
Add first pass at v1.8.0 l200 -> permissioned upgrade script

ajsutton committed:
Deploy new mips.
Set AnchorStateRegistry blueprint address.

ajsutton committed:
Load all contract implementation addresses from the latest contracts version (even if unchanged from earlier versions).

ajsutton committed:
Reuse existing Mips deployment.

ajsutton committed:
Add MIPS address to standard-addresses.json

ajsutton committed:
Fix shellcheck warnings.

ajsutton committed:
Fix more shellcheck warnings.

ajsutton committed:
Fix forge fmt and print errors to stderr.

ajsutton commented on packages/contracts-bedrock/scripts/upgrades/v1.8.0-permissioned/scripts/common.sh:
```diff
@@ -0,0 +1,234 @@
+# Used by fetch_standard_address and fetch_superchain_config_address
+declare -A CACHED_TOML_CONTENT
+
+export CONTRACTS_VERSION=${CONTRACTS_VERSION:-1.8.0-rc.3}
```
Will need to adjust this to the final version once it's available and check if we need different versions for sepolia and mainnet.

ajsutton committed:
Pass contracts-version as an arg when building docker image.

ajsutton committed:
Preserve gas paying token in SystemConfig

ajsutton committed:
Remove incorrect comma.

ajsutton committed:
Fix SystemConfig.initialize types

ajsutton committed:
More initialize adjustments.

Speed up rebuilds

ajsutton commented:
Simulated execution of upgrade on sepolia devnet: https://dashboard.tenderly.co/public/safe/safe-apps/simulator/682a6103-b284-427b-8b80-a4ba29f02f75/state-diff
It's not a particularly realistic case as it's already running permissionless proofs but useful to check things. There's an unexpected change in SystemConfig though with key `0x0000000000000000000000000000000000000000000000000000000000000066` being zeroed. That's the scalar which is complicated by an unreleased Ecotone change to the contracts which is now shipping with Holocene.  I'm guessing the upgrade didn't reinitialise SystemConfig and it's not possible anymore to reinitialise it without changing the scalar. It may be better to not try and reinitialise since we're making the changes we need via StorageSetter anyway.

Inphi commented on packages/contracts-bedrock/scripts/upgrades/v1.8.0-permissioned/DeployUpgrade.s.sol:
```diff
@@ -0,0 +1,404 @@
+        // Run standard assertions.
+        ChainAssertions.checkDisputeGameFactory(contracts, systemOwnerSafe, true);
+        ChainAssertions.checkPermissionedDelayedWETH(contracts, cfg, true, systemOwnerSafe);
+        ChainAssertions.checkOptimismPortal2(contracts, cfg, false);
```
Worth also checking the expected game type for `CANNON_PERMISSIONED`. `ChainAssertions` doesn't do this.

ajsutton replied:
Good call, added.

Inphi reviewed (commented):
The `DeployUpgrade` script looks good to me.

Could you add a README explaining how to use the upgrade scripts.

tynes commented on packages/contracts-bedrock/scripts/upgrades/v1.8.0-permissioned/scripts/bundle.sh:
```diff
@@ -0,0 +1,672 @@
+# We need to re-generate the SystemConfig initialization call
+# We want to use the exact same values that the SystemConfig is already using
+SYSTEM_CONFIG_OWNER=$(cast call "$SYSTEM_CONFIG_PROXY" "owner()")
+SYSTEM_CONFIG_BASE_FEE_SCALAR=$(cast call "$SYSTEM_CONFIG_PROXY" "basefeeScalar()")
```
This is where the bug is. I believe that you need to call `scalar()` and then assert its on version 1 (most significant byte is `0x01`) then decode it with the inverse of https://github.com/ethereum-optimism/optimism/blob/34d5f66ade24bd1f3ce4ce7c0a6cfc1a6540eca1/packages/contracts-bedrock/src/L1/SystemConfig.sol#L406

The logic is also [here](https://github.com/ethereum-optimism/optimism/tree/develop/op-chain-ops/cmd/ecotone-scalar) for decoding

tynes replied:
```bash
$ cast call 0xa6b72407e2dc9EBF84b839B69A24C88929cf20F7 "basefeeScalar()"
0x0000000000000000000000000000000000000000000000000000000000000000
```

Can confirm it does return 0

```bash
$ cast call 0xa6b72407e2dc9EBF84b839B69A24C88929cf20F7 "scalar()"
0x010000000000000000000000000000000000000000000000000d273000001db0
```

this does not return 0, this is the scalar value

tynes replied:
This is a consistency problem due to new storage slots being introduced for the basefee scalar rather than it being a function that modifies the single source of truth of the scalar

see: https://github.com/ethereum-optimism/optimism/blob/34d5f66ade24bd1f3ce4ce7c0a6cfc1a6540eca1/packages/contracts-bedrock/src/L1/SystemConfig.sol#L403-L404

This probably should have been implemented in a way that didn't add new storage slots to prevent this consistency issue

ajsutton replied:
I've removed the re-initialization of SystemConfig from this task as it assumes that the chain is already running `op-contracts/v1.8.0` and the StorageSetter is already being used to zero out the L2OO address and set the DisputeGameFactory address so the initialisation would be a no-op apart from this scalar issue.

How to handle the scalar is being tracked in https://github.com/ethereum-optimism/superchain-ops/issues/394 but that is part of the upgrade to op-contracts/v1.8.0 and this runbook is starting from already running 1.8.0 (just in L2OO config but the SystemConfig should already have been upgraded).

ajsutton committed:
deploy-script: Verify permissioned game impl is set correctly.

ajsutton committed:
Add readme that links to runbook.

ajsutton commented on packages/contracts-bedrock/scripts/upgrades/v1.8.0-permissioned/README.md:
```diff
@@ -0,0 +1,15 @@
+## Usage
+
+For full instructions, including the required preparation and infrastructure changes to perform this upgrade, refer to
+the [full runbook](https://github.com/ethereum-optimism/superchain-ops/blob/a8a3f2a1aeda8d916081be3f4e2a8526286faa4d/runbooks/1.8.0-l2oo-to-permissioned-fps.md).
```
We'll need to update this link one the runbook is merged.

ajsutton committed:
Skip re-initializing the SystemConfig.

There's no simple way to re-init with the same values because of the way scalar is handled in setEcotoneGasConfig

ajsutton committed:
Add note that SystemConfig version may not be changed if already running the correct version.

ajsutton committed:
Do not expect SystemConfig impl address to change.

We aren't re-initializing so it's important it doesn't change versions. If it does for some reason that needs to be properly investigated.

ajsutton committed:
Remove unused variable.

geoknee commented on packages/contracts-bedrock/scripts/upgrades/v1.8.0-permissioned/scripts/common.sh:
```diff
@@ -0,0 +1,232 @@
+
+    # Find the section for required contracts release
+    local section_content
+    section_content=$(echo "$toml_content" | awk -v version="$release_version" '
+        $0 ~ "^\\[releases.\"op-contracts/v" version "\"\\]" {
```
We recently did some clean up in the registry to un-nest the tables from under "releases" https://github.com/ethereum-optimism/superchain-registry/pull/768. 

You'll need a fix equivalent to this https://github.com/ethereum-optimism/optimism/pull/13412/commits/0e054582b9a961be958d54d3653dff7df8f6e235

ajsutton replied:
We're using a fixed commit of the superchain registry in this case anyway. There's a philosophical debate about whether that's the right thing or not but it's what other scripts like this have done and OPCM is ultimately the philosophically correct answer.

opgitgovernance commented:
This pr has been automatically marked as stale and will be closed in 5 days if no updates

pauldowman commented:
@sbvegan do you know who should take this? I think it's off @ajsutton's plate. 

Related:
-  https://github.com/ethereum-optimism/optimism/issues/12990
- https://github.com/ethereum-optimism/superchain-ops/pull/393

sbvegan commented:
@pauldowman I put the issue you linked in the solution's team backlog. Idk who's going to take it at the moment.

ajsutton commented:
Ah @pauldowman and I were both cleaning up GitHub... Doesn't seem like we need this at the moment so closing for now.
